bundles/org.openhab.binding.tibber/src/main/java/org/openhab/binding/tibber/internal/handler/TibberHandler.java

```diff
@@ -0,0 +1,503 @@
+        }
+        if (isConnected()) {
+            close();
+        }
+        super.dispose();
+    }
+
+    public void open() throws Exception {
+        if (isConnected()) {
+            logger.warn("Open: connection is already open");
+        }
+        logger.info("Connecting to: {}", SUBSCRIPTION_URL);
+
+        sslContextFactory = new SslContextFactory(true);
+        sslContextFactory.setTrustAll(true);
+        sslContextFactory.setEndpointIdentificationAlgorithm(null);
+
+        client = new WebSocketClient(sslContextFactory);
+        client.setMaxIdleTimeout(360 * 1000);
+
+        socket = new TibberWebSocketListener();
+
+        request = new ClientUpgradeRequest();
+        String token = new StringBuilder("Bearer ").append(configuration.getToken()).toString();
+        request.setHeader("Authorization", token);
+        request.setSubProtocols("graphql-subscriptions");
+
+        client.start();
+        client.connect(socket, new URI(SUBSCRIPTION_URL), request);
+    }
+
+    public void close() {
+        if (session != null) {
+            String disconnect = "{\"type\":\"connection_terminate\",\"payload\":null}";
+            try {
+                socket.sendMessage(disconnect);
+            } catch (IOException e) {
+                logger.error("Closing websocket stream");
+            }
+            session.close();
+            session = null;
+        }
+    }
+
+    public boolean isConnected() {
+        if (session == null || !session.isOpen()) {
+            return false;
+        }
+        return connected;
+    }
+
+    @WebSocket
+    public class TibberWebSocketListener {
+
+        @OnWebSocketConnect
+        public void onConnect(Session wssession) {
+            session = wssession;
+            connected = true;
+            logger.info("Connected to Server");
+
+            String connection = "{\"type\":\"connection_init\", \"payload\":\"token=" + configuration.getToken()
+                    + "\"}";
+            try {
+                sendMessage(connection);
+            } catch (IOException e) {
+                logger.error("Send Message Exception: ", e);
+            }
+        }
+
+        @OnWebSocketClose
+        public void onClose(int statusCode, String reason) {
+            logger.warn("Closing a WebSocket due to {}", reason);
+            session = null;
+            connected = false;
+        }
+
+        @OnWebSocketError
+        public void onWebSocketError(Throwable e) {
+            logger.error("Error during websocket communication: {}", e.getMessage(), e);
+            onClose(0, e.getMessage());
+        }
+
+        @OnWebSocketMessage
+        public void onMessage(String message) {
+            if (message.contains("connection_ack")) {
+                startSubscription();
+            } else if (message.contains("error")) {
+                close();
+                logger.warn("WebSocket Connection closed due to Connection error: {}", message);
```

Comment:
Well, not really. The API is set up both for http requests and live websocket stream. Even if the websocket connection is closed, the thing could and should remain ONLINE for http request/response. If "error" is received as part of http request, this will trigger state change to OFFLINE. As for the websocket connection, it is included as part of the "Refresh" method (schedule) to try to reestablish connection every 1 minute.